### PR TITLE
Adding analog sonar driver. Closes #7664

### LINKED ISSUE #7664: Maxbotix sonar sensor not working
Hi,

I would like to use maxbotix MB1240 sonar sensor with px4, but unfortunately it's not working.

I've connected the sonar according to [this](http://ardupilot.org/copter/docs/common-rangefinder-maxbotix-analog.html) guide. I hear the sonar, so it should work. Then I uncommented line **#drivers/mb12xx** in **nuttx_px4fmu-v2_default.cmake** file. After that I just compiled the whole thing with `make px4fmu-v2_default` command and uploaded it with `make px4fmu-v2_default upload`. Finally I enabled the **SENS_EN_MB12XX** parameter.

But when I try to display the **distance_sensor** message in MAVLink inspector, I can't see it.

I'm using px4 firmware v1.6.5 with QGroundControl v3.2.1.

Does anybody know what's going on? Did I something wrong or am I missing something?

Thank you for answer.

### PR DESCRIPTION
**Pull request by L4ncelot**
Adding driver for analog sonar sensor. Tested on maxbotix mb 1240 sensor.

### CODE REVIEW

**Comment by TSC21** [comment]
Can you show us the output of `DISTANCE_SENSOR` on QGC for this?

**Comment by TSC21** [comment]
Also can you run `make format` to fix style?

**Comment by L4ncelot** [comment]
Here are the screenshots:
screenshot from 2017-07-27 17-19-03
screenshot from 2017-07-27 17-20-10

Btw, I have again the same problem with the squash... If I do git `rebase -i HEAD~2` I can see only my commits, but it's rebasing then to another branch...

**Comment by TSC21** [comment]
> Btw, I have again the same problem with the squash... If I do git rebase -i HEAD~2 I can see only my commits, but it's rebasing then to another branch...

But aren't you writing code over master? The idea is that you clone master, create a branch, write code over it, commit, pull request.

**Comment by L4ncelot** [comment]
Here's what I did. I forked the master to my repository, cloned the repository, created new branch, added the changes, commited the changes, pulled the master to resolve conflicts and pushed to the repository.

**Comment by L4ncelot** [comment]
Here's the output from `git rebase -i HEAD~8`:
```
pick c44322c MAVLink app: Use more advanced forwarding logic
pick 9124617 MAVLink: Always enable forwarding for companion link
pick 7612b94 Allow relaying from same system ID or with target component 0
pick 1eda66c Code format
pick dc8caea MAVLink: Default to standard stream config
pick 63e8234 Adding analog sonar driver.
pick cd06e05 Adding analog sonar driver.
pick 4a55747 Reformating the code.

```

But I can see only 3 commit related to this PR, not the 4th one.

**Comment by TSC21** [comment]
> pulled the master to resolve conflicts and pushed to the repository.

Why did you do that again if you already had forked master? Was it the current master? Why don't you do like this: clone "your" master, `git pull origin master` and `git push` to your master again. Then, from "your" master, `git branch -b analog-sonar-driver2`, make changes, commit changes, `git push -u <your-fork> analog-sonar-driver2` and then do a PR.

**Comment by TSC21** [comment]
> But I can see only 3 commit related to this PR, not the 4th one.

Ok do like this:
`git rebase -i HEAD~4`
And then:
```bash
pick dc8caea MAVLink: Default to standard stream config
pick 63e8234 Adding analog sonar driver.
fixup cd06e05 Adding analog sonar driver.
fixup 4a55747 Reformating the code.
```
`fixup` acts like squash but discards the commit log message.
The you can do `git push -u <your-fork> analog-sonar-driver --force`.

**Comment by L4ncelot** [comment]
Well I guess I did it in the wrong order. I pulled the master to resolve the conflicts, but I haven't thought about doing it before creating the branch... sorry for complications. I'll try the fixup method.

File: src/modules/sensors/sensors.cpp [hunk 1]
```diff
@@ -166,13 +167,15 @@ class Sensors : public ModuleBase<Sensors>
 	int		_diff_pres_sub;			/**< raw differential pressure subscription */
 	int		_vcontrol_mode_sub;		/**< vehicle control mode subscription */
 	int 		_params_sub;			/**< notification of parameter updates */
+	int     _orb_class_instance;
```
Comment (TSC21) [inline src/modules/sensors/sensors.cpp]: Also can you please fix the format here and add the description comment?

File: src/modules/sensors/sensors.cpp [hunk 2]
```diff
@@ -254,6 +261,10 @@ Sensors::Sensors(bool hil_enabled) :
 	memset(&_diff_pres, 0, sizeof(_diff_pres));
 	memset(&_parameters, 0, sizeof(_parameters));
 
+	_orb_class_instance = -1;
+	_min_sonar_distance = 0.2f;
+	_max_sonar_distance = 7.65f;
```
Comment (TSC21) [inline src/modules/sensors/sensors.cpp]: The above should be handled as configurable PX4 parameters (as the sonars may also be different)

File: src/modules/sensors/sensors.cpp [hunk 3]
```diff
@@ -490,71 +501,102 @@ Sensors::adc_poll(struct sensor_combined_s &raw)
...
+
+					if (ADC_SONAR_VOLTAGE_CHANNEL == buf_adc[i].am_channel) {
+#endif
+				float voltage_scaling = 2.04f;
+				float distance = buf_adc[i].am_data * 0.001f * voltage_scaling;
+
+				struct distance_sensor_s distance_report;
+				distance_report.timestamp = t;
+				distance_report.type = distance_sensor_s::MAV_DISTANCE_SENSOR_ULTRASOUND;
+				distance_report.orientation = 8;
```
Comment (TSC21) [inline src/modules/sensors/sensors.cpp]: Again, shouldn't one be able to choose what orientation and sensor type is being imputed?

File: src/modules/sensors/sensors.cpp [hunk 4]
```diff
@@ -490,71 +501,102 @@ Sensors::adc_poll(struct sensor_combined_s &raw)
...
+				distance_report.current_distance = distance;
+				distance_report.min_distance = _min_sonar_distance;
+				distance_report.max_distance = _max_sonar_distance;
+				distance_report.covariance = 0.0f;
```
Comment (TSC21) [inline src/modules/sensors/sensors.cpp]: Does the sensor datasheet specify a variance for sensor readings?

Comment (L4ncelot) [inline src/modules/sensors/sensors.cpp]: Well [here](https://www.maxbotix.com/documents/XL-MaxSonar-EZ_Datasheet.pdf) is the datasheet, but I couldn't find this parameter in it.

File: src/modules/sensors/sensors.cpp (hunk 3 above)
Comment (L4ncelot) [inline src/modules/sensors/sensors.cpp]: Yes, you're right. I've made to just work for my own example. I'll edit it tomorrow.

File: src/drivers/boards/px4fmu-v2/board_config.h [hunk 5]
```diff
@@ -159,6 +159,7 @@
 #define ADC_BATTERY_CURRENT_CHANNEL	3
 #define ADC_5V_RAIL_SENSE		4
 #define ADC_AIRSPEED_VOLTAGE_CHANNEL	15
+#define ADC_SONAR_VOLTAGE_CHANNEL       14
```
Comment (TSC21) [inline src/drivers/boards/px4fmu-v2/board_config.h]: Also, add this to the other board configs that have the ADC available.

File: src/modules/sensors/sensors.cpp [hunk 6]
```diff
@@ -166,13 +167,15 @@ class Sensors : public ModuleBase<Sensors>
...
 	int 		_params_sub;			/**< notification of parameter updates */
+	int     _orb_class_instance;
 
 	orb_advert_t	_sensor_pub;			/**< combined sensor data topic */
 	orb_advert_t	_battery_pub[BOARD_NUMBER_BRICKS] = {nullptr};			/**< battery status */
 	int 			_battery_pub_intance0ndx = 0; /**< track the index of instance 0 */
 	orb_advert_t	_airspeed_pub;			/**< airspeed */
 	orb_advert_t	_diff_pres_pub;			/**< differential_pressure */
 	orb_advert_t	_sensor_preflight;		/**< sensor preflight topic */
+	orb_advert_t    _distance_sensor_topic;
```
Comment (TSC21) [inline src/modules/sensors/sensors.cpp]: comment

**Comment by TSC21** [comment]
Just an interesting thing to have, but not a need, wish you can find on the datasheet:
> **Temperature Compensation**
The speed of sound in air increases about 0.6 meters per second, per degree centigrade. The XL-MaxSonar-EZ sensors are not equipped with an internal temperature compensation. If temperature compensation is desired, contact MaxBotix and request the temperature compensation formula PDF. This will allow users to compensation for speed of sound changes.

Could be cool if you could add this for both the analog and the I2C driver.

File: src/modules/sensors/sensors.cpp (hunk 3 above)
Comment (L4ncelot) [inline src/modules/sensors/sensors.cpp]: I'm trying to set the distance sensor type as a parameter and I would like to make it so user can choose the type from drop down menu (e.g. like `SYS_MC_EST_GROUP`). 
This is what has been generated for me in `parameters.xml` file:
```
<parameter default="0" name="DIST_SEN_TYPE" type="INT32">
      <short_desc>Distance sensor type</short_desc>
      <min>0</min>
      <max>3</max>
      <reboot_required>false</reboot_required>
      <scope>modules/sensors</scope>
      <values>
        <value code="1">ULTRASOUND</value>
        <value code="0">LASER</value>
        <value code="3">RADAR</value>
        <value code="2">INFRARED</value>
      </values>
    </parameter>
```
But I can't see the dropdown menu. I can only set the value manually. I also did `(void)param_find("DIST_SEN_TYPE");` according to this comment: "These are parameters for which QGroundControl always expects to be returned in a list request." in `modules/sensors/parameters.cpp` file. 

Am I missing something else?

**Comment by TSC21** [comment]
@L4ncelot updates?

**Comment by L4ncelot** [comment]
@TSC21 Well I had some problem with the pop down menu in QGroundControl as described above and couldn't figure out where the problem was. Since then I had to work on something else. But besides that it's ready to be merge I guess. But after final review I'll probably have to do last PR, because of my stupid pull of master branch :).

**Comment by TSC21** [comment]
> But besides that it's ready to be merge I guess.

There are some things above to check (code style, comments, indentation).

> @TSC21 Well I had some problem with the pop down menu in QGroundControl as described above and couldn't figure out where the problem was.

Try to reboot the board. Also, how are you generating the parameters?

**Comment by TSC21** [comment]
Also you need to add your parameter to a group (example: https://github.com/PX4/Firmware/blob/c4638c3da45d8d10514ef6a321774b55a9af098d/src/modules/local_position_estimator/params.c#L19)

**Comment by TSC21** [comment]
@L4ncelot updates?

**Comment by dagar** [comment]
Closing as stale. Please rebase on master and open a new pull request.

**Comment by kyuhyong** [comment]
I am also trying to enable analog sonar input as vertical ranging for terrain following. 
How can I merge this specific branch to my cloned master? Or which branch should I check out?
Thanks.

**Pull request closed without merging**